models/yasgbt.py

```diff
@@ -0,0 +1,90 @@
+import numpy as np
+from tqdm.notebook import tqdm
+
+from sklearn.base import BaseEstimator
+from sklearn.ensemble import HistGradientBoostingRegressor
+
+from sksurv.functions import StepFunction
+from sksurv.nonparametric import kaplan_meier_estimator, CensoringDistributionEstimator
+
+from models.survival_mixin import SurvivalMixin
+
+
+class DataSampler:
+    
+    def __init__(self, y):
+        self.y = y
+        times, km = kaplan_meier_estimator(y["event"], y["duration"])
+        # inverse_km has x: [0.01, ..., .99] and y: [0., ..., 820.]
+        self.inverse_km = StepFunction(x=km[::-1], y=times[::-1])
+        self.max_km, self.min_km = km[0], km[-1]
+    
+    def make_sample(self):
+        y = self.y
+        
+        surv_probs = np.random.uniform(self.min_km, self.max_km, y.shape[0])
+        times = self.inverse_km(surv_probs)
+
+        mask_y_0 = y["event"] & (y["duration"] <= times)
+        mask_y_1 = y["duration"] > times
+
+        yc = np.zeros(y.shape[0])
+        yc[mask_y_1] = 1
+        
+        cens = CensoringDistributionEstimator().fit(y)
+        # calculate inverse probability of censoring weight at current time point t.
+        prob_cens_t = cens.predict_proba(times)
+        prob_cens_t[prob_cens_t == 0] = np.inf
+        # calculate inverse probability of censoring weights at observed time point
+        prob_cens_y = cens.predict_proba(y["duration"])
+        prob_cens_y[prob_cens_y == 0] = np.inf
+
+        sample_weights = np.where(mask_y_0, prob_cens_y, 0)
+        sample_weights = np.where(mask_y_1, prob_cens_t, sample_weights)
+                
```

Comment:
I think we should enforce the following to avoid introducing non-zero terms in the loss because of the censoring pattern:

```python
    # Ensure that all other terms (censored data points with a duration below the
    # sampled time) contribute nothing to the loss. The value of the classification
    # target y_c is undefined for those:
    remaining_terms = np.logical_not(mask_y_0) & np.logical_not(mask_y_1)
    sample_weights[remaining_terms] = 0
```

The inverse censoring probability weighting alone would be enough to zero out those terms: the ICPW should be between 1 and +inf by construction.

Reply:
Actually this is indeed redundant with what is done in:

```python
        sample_weights = np.where(mask_y_0, prob_cens_y, 0)
        sample_weights = np.where(mask_y_1, prob_cens_t, sample_weights)
```

because of the right hand side `0` in the first where.

Reply:
It might be interesting to add an inline comment to explain that the `y_c` terms that do not match either of `mask_y_0` and `mask_y_1` (censored data points with a duration below the sampled time) have 0 weight as they should not contribute to the time-varying BS loss.